File: protocol-generator/swift/Sources/EntityCoreProtocol/Identity.swift
// Identity.swift — L1 identity bundle: keypair, system/peer, peer_id, signing.
//
// SPEC-FIRST peer_id construction (§1.5, the canonical-form table). For Ed25519
// (key_type 0x01) the canonical hash_type is 0x00 identity-multihash and the
// digest IS the raw 32-byte public key (v7.64/v7.65). The peer_id string is
//   Base58(varint(0x01) || varint(0x00) || public_key).
//
//   ⚑ A-SW-008 (§7.4-vs-§1.5 peer-id contradiction — see SPEC-AMBIGUITY-LOG).
//   §7.4 line 4001 lists "Peer ID derivation (§7.4)" in the §9.1 floor and §7.4's
//   prose historically read as SHA-256(pubkey); §1.5's canonical-form table (v7.65)
//   mandates identity-multihash (digest = raw pubkey) for Ed25519. The v7.74 §7.4
//   text now DEFERS to the §1.5 table ("See §1.5 canonical-form per key_type"), so
//   the contradiction is *narrowed* but the floor bullet still cites §7.4 by number.
//   We construct per §1.5 (identity-multihash) — the canonical, cap-pattern-bearing
//   form. Corroborates OCaml A-OC-007 / Zig A-ZIG-001 (prior peers hit the same).
//
// system/peer entity (§3.5, v7.65): data is {public_key (bytes), key_type (string)}
// — peer_id is NOT in the hashable basis (P×I primitive discipline). The content
// hash of the peer entity is the identity hash used in caps/signatures.

/// A peer's identity: the Ed25519 keypair, the derived peer_id, and the
/// materialized `system/peer` entity (whose content_hash is the identity hash).
public struct Identity: Sendable {
    /// 32-byte Ed25519 seed (private key material).
    public let seed: [UInt8]
    /// 32-byte Ed25519 public key.
    public let publicKey: [UInt8]
    /// Base58 peer_id (§1.5 identity-multihash form for Ed25519).
    public let peerID: String
    /// The materialized `system/peer` entity.
    public let peerEntity: BuiltEntity
    /// content_hash of the peer entity — the identity hash referenced by caps,
    /// signatures, grants (§3.5 / §5.2: grantee/signer/author are all this).
    public var identityHash: [UInt8] { peerEntity.hash }

    public init(seed: [UInt8]) throws(CodecError) {
        guard seed.count == 32 else { throw .badSeed }
        self.seed = seed
        self.publicKey = try Signing.publicKey(fromSeed: seed)
        // §1.5 canonical-form: Ed25519 → identity-multihash, digest = raw pubkey.
        self.peerID = try PeerID.fromEd25519(publicKey: publicKey).format()
        // system/peer hashable basis = {public_key, key_type} (v7.65; no peer_id).
        // key_type is the canonical lowercase-ASCII string (§1.5 line 496): "ed25519".
        self.peerEntity = try Model.make(type: "system/peer", fields: [
            ("public_key", .bytes(publicKey)),
            ("key_type", .text("ed25519")),
        ])
    }

    /// Sign an entity's content_hash (§7.3: message = full content_hash bytes).
    public func sign(contentHash: [UInt8]) throws(CodecError) -> [UInt8] {
        try Signing.sign(seed: seed, message: contentHash)
    }

    /// Build a `system/signature` entity (§3.5) over `target` by this identity.
    /// `signer` = this identity's hash (the peer-entity content_hash).
    public func signatureEntity(target: [UInt8]) throws(CodecError) -> BuiltEntity {
        let sig = try sign(contentHash: target)
        return try Model.make(type: "system/signature", fields: [
            ("target", .bytes(target)),
            ("signer", .bytes(identityHash)),
            ("algorithm", .text("ed25519")),
            ("signature", .bytes(sig)),
        ])
    }
}

File: protocol-generator/forth/src/identity.fs
\ entity-core-protocol-forth — L1 identity: keystore + peer identity + signatures.
\
\ An identity is derived from a 32-byte Ed25519 SEED (the entity-core PEM = base64 of the
\ raw seed). From the seed we derive: the 32-byte public key (FFI), the `system/peer`
\ entity {public_key, key_type} (§3.5 — peer_id is NOT in the hashable basis), the
\ id_hash = content_hash(peer_entity), and the wire peer_id string (§1.5 canonical form:
\ Base58(varint(0x01) ‖ varint(0x00) ‖ raw-pubkey) — hash_type 0x00 identity-multihash).
\
\ Single-threaded peer: one identity per process. We keep the seed + derived material in
\ module VARIABLEs (copied into the durable store heap so they outlive an arena reset).

create id-seed  32 allot          \ the 32-byte seed (durable module storage)
create id-pub   32 allot          \ derived public key
variable id-peer-addr             \ durable addr of the system/peer entity record
variable id-peer-len
variable id-hash-addr             \ durable addr of id_hash (content_hash of peer entity)
variable id-hash-len
variable id-peerid-addr           \ durable addr of the base58 peer_id string
variable id-peerid-len

s" ed25519" 2constant KEY-TYPE-STR
1 constant KEY-TYPE-ED25519       \ §1.5 key_type code 0x01
0 constant HASH-TYPE-IDENTITY     \ §1.5 hash_type 0x00 identity-multihash

\ id-peer-entity ( -- eaddr eu )  build the system/peer entity {public_key, key_type} in
\ the arena. §3.5: exactly these two fields, canonical map order handled by the codec.
: id-peer-entity ( -- eaddr eu )
  am-mark { mk }
  [char] m b,  2 4 >be
  s" public_key" tv-text 2drop  id-pub 32 tv-bytes 2drop
  s" key_type"   tv-text 2drop  KEY-TYPE-STR tv-text 2drop
  mk am-span                                   \ ( data-map-addr data-map-len )
  s" system/peer" 2swap ent-make ;             \ ( type data -> ent )

\ id-init ( seed-addr seed-u -- )  install the identity from a 32-byte seed. Derives pub,
\ peer entity, id_hash, peer_id; stores durable copies. THROWs E-BAD-SEED on wrong length.
: id-init { saddr su -- }
  su 32 <> if E-BAD-SEED throw then
  saddr id-seed 32 move
  id-seed 32 crypto-ed25519-pubkey  drop id-pub 32 move   \ derive pubkey -> id-pub
  id-peer-entity  store-dup { pu } { paddr }              \ durable peer entity
  paddr id-peer-addr !  pu id-peer-len !
  paddr ent-hash store-dup id-hash-len ! id-hash-addr !   \ durable id_hash
  KEY-TYPE-ED25519 HASH-TYPE-IDENTITY id-pub 32 peerid-format
  store-dup id-peerid-len ! id-peerid-addr ! ;            \ durable peer_id string

: id-peer      ( -- eaddr eu )  id-peer-addr @ id-peer-len @ ;
: id-idhash    ( -- h-addr h-u )  id-hash-addr @ id-hash-len @ ;
: id-peerid    ( -- p-addr p-u )  id-peerid-addr @ id-peerid-len @ ;

\ ── signatures (§3.5 system/signature) ──
\ sig-entity ( target-addr target-u signer-addr signer-u sig-addr(64) -- eaddr eu )
\ build a system/signature {target, signer, algorithm:"ed25519", signature} entity.
: sig-entity { taddr tu sgnaddr sgnu sigaddr -- eaddr eu }
  am-mark { mk }
  [char] m b,  4 4 >be
  s" algorithm" tv-text 2drop  s" ed25519" tv-text 2drop
  s" signature" tv-text 2drop  sigaddr 64 tv-bytes 2drop
  s" signer"    tv-text 2drop  sgnaddr sgnu tv-bytes 2drop
  s" target"    tv-text 2drop  taddr tu tv-bytes 2drop
  mk am-span
  s" system/signature" 2swap ent-make ;

\ id-sign ( target-hash-addr target-hash-u -- sig-entity-addr sig-entity-u )  sign the
\ target content-hash bytes with our seed; build the system/signature entity (signer =
\ our id_hash). The full hash bytes (format-code ‖ digest) are the signed message (§4.6).
: id-sign { thaddr thu -- eaddr eu }
  id-seed  thaddr thu  crypto-ed25519-sign  drop { sigaddr }   \ 64-byte sig in arena
  thaddr thu  id-idhash  sigaddr  sig-entity ;

\ id-verify-sig ( sig-ent-addr pub-addr pub-u -- flag )  verify a system/signature entity's
\ signature over its `target` using the given public key. 1 valid / 0 invalid.
: id-verify-sig { sigent paddr pu -- flag }
  paddr pu 32 <> if 2drop drop false exit then drop
  sigent s" target"    ent-field dup 0= if drop false exit then tv-payload { tu } { taddr }
  sigent s" signature" ent-field dup 0= if drop false exit then tv-payload { su } { saddr }
  su 64 <> if false exit then
  paddr taddr tu saddr crypto-ed25519-verify ;

\ id-peerid-of-pub ( pub-addr pub-u -- peerid-addr peerid-u )  derive the §1.5 peer_id
\ string of an Ed25519 public key (identity-multihash form).
: id-peerid-of-pub { paddr pu -- pidaddr pidu }
  KEY-TYPE-ED25519 HASH-TYPE-IDENTITY paddr pu peerid-format ;
